### DISCUSSION TITLE
Token refresh failed: %TeslaApi.Error message: socket closed

### DISCUSSION THREAD

**Discussion by kongomongo** [teslamate-org/teslamate#1294] [Q&A]
**Describe the bug**
Finding this message in log:
```
teslamate_1               | 2021-02-04 09:02:28.256804 waiting for postgres at database:5432
teslamate_1               | 2021-02-04 09:02:30.466437 2021-02-04 08:02:30.463 [info] == Running 20210130174838 TeslaMate.Repo.Migrations.UseVariableLengthTypeForTokens.change/0 forward
teslamate_1               | 2021-02-04 09:02:30.466775 2021-02-04 08:02:30.466 [info] alter table tokens
teslamate_1               | 2021-02-04 09:02:30.493866 2021-02-04 08:02:30.493 [info] == Migrated 20210130174838 in 0.0s
teslamate_1               | 2021-02-04 09:02:32.497925 2021-02-04 08:02:32.497 [info] Version: 1.21.2
teslamate_1               | 2021-02-04 09:02:33.127921 2021-02-04 08:02:33.127 [warn] Token refresh failed: %TeslaApi.Error{env: nil, message: "socket closed", reason: :token_refresh}
teslamate_1               | 2021-02-04 09:02:33.127967 2021-02-04 08:02:33.127 [info] Scheduling token refresh in 1 h
teslamate_1               | 2021-02-04 09:02:33.130506 2021-02-04 08:02:33.130 [info] Running TeslaMateWeb.Endpoint with cowboy 2.8.0 at :::4000 (http)
teslamate_1               | 2021-02-04 09:02:33.136694 2021-02-04 08:02:33.136 [info] Access TeslaMateWeb.Endpoint at http://localhost
teslamate_1               | 2021-02-04 09:02:35.872012 2021-02-04 08:02:35.871 [info] Starting logger for 'xxx'
teslamate_1               | 2021-02-04 09:02:36.571985 2021-02-04 08:02:36.571 car_id=1 [info] Start / :online
teslamate_1               | 2021-02-04 09:02:36.596009 2021-02-04 08:02:36.595 car_id=1 [info] Connecting ...
teslamate_1               | 2021-02-04 09:02:39.523960 2021-02-04 08:02:39.523 [info] tzdata release in place is from a file last modified Tue, 22 Dec 2020 23:35:21 GMT. Release file on server was last modified Sun, 24 Jan 2021 19:35:23 GMT.
teslamate_1               | 2021-02-04 09:02:41.038489 2021-02-04 08:02:41.037 [info] Tzdata has updated the release from 2020e to 2021a
teslamate_1               | 2021-02-04 09:05:40.522382 2021-02-04 08:05:40.521 car_id=1 [info] Suspending logging
teslamate_1               | 2021-02-04 09:31:23.738424 2021-02-04 08:31:23.737 car_id=1 [info] Driving / Start
teslamate_1               | 2021-02-04 10:00:01.946375 2021-02-04 09:00:01.945 car_id=1 [info] Fetch already in progress ...
teslamate_1               | 2021-02-04 10:00:05.293459 2021-02-04 09:00:05.292 car_id=1 [info] Driving / Ended / 25 km – 29 min
teslamate_1               | 2021-02-04 10:00:05.293507 2021-02-04 09:00:05.292 car_id=1 [info] Start / :online
teslamate_1               | 2021-02-04 10:02:58.815695 2021-02-04 09:02:58.815 car_id=1 [info] Driving / Start
teslamate_1               | 2021-02-04 10:03:02.129666 2021-02-04 09:03:02.127 [info] Refreshing access token ...
teslamate_1               | 2021-02-04 10:03:02.795853 2021-02-04 09:03:02.794 [error] GenServer TeslaMate.Api terminating
teslamate_1               | 2021-02-04 10:03:02.795890 ** (MatchError) no match of right hand side value: {:error, %TeslaApi.Error{env: nil, message: "socket closed", reason: :token_refresh}}
teslamate_1               | 2021-02-04 10:03:02.795896     (teslamate 1.21.2) lib/teslamate/api.ex:150: TeslaMate.Api.handle_info/2
teslamate_1               | 2021-02-04 10:03:02.795900     (stdlib 3.14) gen_server.erl:689: :gen_server.try_dispatch/4
teslamate_1               | 2021-02-04 10:03:02.795904     (stdlib 3.14) gen_server.erl:765: :gen_server.handle_msg/6
teslamate_1               | 2021-02-04 10:03:02.795908     (stdlib 3.14) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
teslamate_1               | 2021-02-04 10:03:02.795912 Last message: :refresh_auth
teslamate_1               | 2021-02-04 10:03:03.030732 2021-02-04 09:03:03.030 [warn] Token refresh failed: %TeslaApi.Error{env: nil, message: "socket closed", reason: :token_refresh}
teslamate_1               | 2021-02-04 10:03:03.030827 2021-02-04 09:03:03.030 [info] Scheduling token refresh in 1 h
teslamate_1               | 2021-02-04 10:31:21.369940 2021-02-04 09:31:21.369 car_id=1 [info] Driving / Ended / 25 km – 28 min
teslamate_1               | 2021-02-04 10:31:21.369999 2021-02-04 09:31:21.369 car_id=1 [info] Start / :online
teslamate_1               | 2021-02-04 10:34:39.792039 2021-02-04 09:34:39.791 car_id=1 [info] Suspending logging
teslamate_1               | 2021-02-04 10:36:35.497912 2021-02-04 09:36:35.497 [info] Connection was closed (a:1|t:1|d:1). Reconnecting …
teslamate_1               | 2021-02-04 10:36:47.087725 2021-02-04 09:36:47.087 [info] Connection was closed (a:1|t:2|d:1). Reconnecting …
teslamate_1               | 2021-02-04 10:41:35.301561 2021-02-04 09:41:35.299 [info] Connection was closed (a:1|t:1|d:1). Reconnecting …
teslamate_1               | 2021-02-04 10:41:46.142798 2021-02-04 09:41:46.142 [info] Connection was closed (a:1|t:2|d:1). Reconnecting …

```

**Expected behavior**
No error

**How to reproduce it (as minimally and precisely as possible)**:
Start  teslamate (latest stable)

**Environment**

- TeslaMate version 1.21.2:
- Type of installation: Docker
- OS TeslaMate is installed on: linux

**Comment by kongomongo** [comment]
using :4000 and going to settings -> logging out and relogging in I get:

```
teslamate_1               | 2021-02-04 10:48:17.165993 2021-02-04 09:48:17.165 [info] GET /
teslamate_1               | 2021-02-04 10:48:17.184814 2021-02-04 09:48:17.184 [info] Sent 200 in 18ms
teslamate_1               | 2021-02-04 10:48:25.392609 2021-02-04 09:48:25.392 [info] GET /settings
teslamate_1               | 2021-02-04 10:48:25.402068 2021-02-04 09:48:25.401 [info] Sent 200 in 8ms
teslamate_1               | 2021-02-04 10:48:36.313954 2021-02-04 09:48:36.313 [info] GET /
teslamate_1               | 2021-02-04 10:48:36.316497 2021-02-04 09:48:36.315 [info] Sent 302 in 2ms
teslamate_1               | 2021-02-04 10:48:36.329877 2021-02-04 09:48:36.327 [info] GET /sign_in
teslamate_1               | 2021-02-04 10:48:36.331733 2021-02-04 09:48:36.330 [info] Sent 200 in 3ms
teslamate_1               | 2021-02-04 10:48:56.091904 2021-02-04 09:48:56.091 [info] Scheduling token refresh in 5 wk 5 d
teslamate_1               | 2021-02-04 10:48:56.369344 2021-02-04 09:48:56.368 [info] GET /
teslamate_1               | 2021-02-04 10:49:05.484711 2021-02-04 09:49:05.483 [warn] TeslaApi.Error / %TeslaApi.Error{
teslamate_1               | 2021-02-04 10:49:05.485111   env: %Tesla.Env{
teslamate_1               | 2021-02-04 10:49:05.485118     __client__: %Tesla.Client{adapter: nil, fun: nil, post: [], pre: []},
teslamate_1               | 2021-02-04 10:49:05.485123     __module__: TeslaApi,
teslamate_1               | 2021-02-04 10:49:05.485126     body: %{
teslamate_1               | 2021-02-04 10:49:05.485131       "error" => "vehicle unavailable: {:error=>\"{\\\"error\\\": \\\"timeout\\\"}\"}",
teslamate_1               | 2021-02-04 10:49:05.485143       "error_description" => "",
teslamate_1               | 2021-02-04 10:49:05.485148       "response" => nil
teslamate_1               | 2021-02-04 10:49:05.485152     },
teslamate_1               | 2021-02-04 10:49:05.485156     headers: [
teslamate_1               | 2021-02-04 10:49:05.485160       {"date", "Thu, 04 Feb 2021 09:49:05 GMT"},
teslamate_1               | 2021-02-04 10:49:05.485164       {"content-type", "application/json; charset=utf-8"},
teslamate_1               | 2021-02-04 10:49:05.485184       {"content-length", "118"},
teslamate_1               | 2021-02-04 10:49:05.485188       {"connection", "keep-alive"},
teslamate_1               | 2021-02-04 10:49:05.485192       {"x-frame-options", "SAMEORIGIN"},
teslamate_1               | 2021-02-04 10:49:05.485195       {"x-xss-protection", "1; mode=block"},
teslamate_1               | 2021-02-04 10:49:05.485199       {"x-content-type-options", "nosniff"},
teslamate_1               | 2021-02-04 10:49:05.485202       {"cache-control", "no-cache"},
teslamate_1               | 2021-02-04 10:49:05.485206       {"x-request-id",
teslamate_1               | 2021-02-04 10:49:05.485209        "ccde975620b01fb97392c61c912842f7ccde97xxx0b01fb97392c61c912842f7"},
teslamate_1               | 2021-02-04 10:49:05.485213       {"x-runtime", "9.051839"},
teslamate_1               | 2021-02-04 10:49:05.485216       {"strict-transport-security", "max-age=15724800; includeSubDomains"}
teslamate_1               | 2021-02-04 10:49:05.485220     ],
teslamate_1               | 2021-02-04 10:49:05.485223     method: :get,
teslamate_1               | 2021-02-04 10:49:05.485227     opts: [
teslamate_1               | 2021-02-04 10:49:05.485230       access_token: "qts-exxxx1e68305b38a843120xxxxxxxxxxxf8cd538d92317107c4ecf7b2xxxx3c"
teslamate_1               | 2021-02-04 10:49:05.485235     ],
teslamate_1               | 2021-02-04 10:49:05.485239     query: [],
teslamate_1               | 2021-02-04 10:49:05.485242     status: 408,
teslamate_1               | 2021-02-04 10:49:05.485245     url: "https://owner-api.teslamotors.com/api/1/vehicles/3832xxxxxxxxx2785/vehicle_data"
teslamate_1               | 2021-02-04 10:49:05.485249   },
teslamate_1               | 2021-02-04 10:49:05.485252   message: nil,
teslamate_1               | 2021-02-04 10:49:05.485256   reason: :vehicle_unavailable
teslamate_1               | 2021-02-04 10:49:05.485259 }
teslamate_1               | 2021-02-04 10:49:05.485262 2021-02-04 09:49:05.483 [info] Starting logger for 'xxx'
teslamate_1               | 2021-02-04 10:49:05.492438 2021-02-04 09:49:05.491 [info] Sent 200 in 9123ms
teslamate_1               | 2021-02-04 10:49:14.608529 2021-02-04 09:49:14.607 car_id=1 [info] Start / :online
teslamate_1               | 2021-02-04 10:49:14.617152 2021-02-04 09:49:14.616 car_id=1 [info] Connecting ...
teslamate_1               | 2021-02-04 10:49:26.947338 2021-02-04 09:49:26.946 [info] Connection was closed (a:1|t:1|d:0). Reconnecting …
```

**Comment by dkropman-Slalom** [comment]
I wonder if this is due to the new auth that tesla is doing? https://www.reddit.com/r/TeslaLounge/comments/lbov9o/tesla_owners_api_changed_to_oauth_v3/

Edit - Also looks like this has been fixed in latest release (new oauth)

**Comment by kongomongo** [comment]
I thought 1.21.2 was the latest version.

**Comment by dkropman-Slalom** [comment]
Sorry, my bad, I was updating my comment to say my message is probably not relevant to your error. You have the latest with the new oath code in.

**Comment by adriankumpf** [comment]
> using :4000 and going to settings -> logging out and relogging in I get:
> 
> ```
> teslamate_1               | 2021-02-04 10:48:17.165993 2021-02-04 09:48:17.165 [info] GET /
> teslamate_1               | 2021-02-04 10:48:17.184814 2021-02-04 09:48:17.184 [info] Sent 200 in 18ms
> teslamate_1               | 2021-02-04 10:48:25.392609 2021-02-04 09:48:25.392 [info] GET /settings
> teslamate_1               | 2021-02-04 10:48:25.402068 2021-02-04 09:48:25.401 [info] Sent 200 in 8ms
> teslamate_1               | 2021-02-04 10:48:36.313954 2021-02-04 09:48:36.313 [info] GET /
> teslamate_1               | 2021-02-04 10:48:36.316497 2021-02-04 09:48:36.315 [info] Sent 302 in 2ms
> teslamate_1               | 2021-02-04 10:48:36.329877 2021-02-04 09:48:36.327 [info] GET /sign_in
> teslamate_1               | 2021-02-04 10:48:36.331733 2021-02-04 09:48:36.330 [info] Sent 200 in 3ms
> teslamate_1               | 2021-02-04 10:48:56.091904 2021-02-04 09:48:56.091 [info] Scheduling token refresh in 5 wk 5 d
> teslamate_1               | 2021-02-04 10:48:56.369344 2021-02-04 09:48:56.368 [info] GET /
> teslamate_1               | 2021-02-04 10:49:05.484711 2021-02-04 09:49:05.483 [warn] TeslaApi.Error / %TeslaApi.Error{
> teslamate_1               | 2021-02-04 10:49:05.485111   env: %Tesla.Env{
> teslamate_1               | 2021-02-04 10:49:05.485118     __client__: %Tesla.Client{adapter: nil, fun: nil, post: [], pre: []},
> teslamate_1               | 2021-02-04 10:49:05.485123     __module__: TeslaApi,
> teslamate_1               | 2021-02-04 10:49:05.485126     body: %{
> teslamate_1               | 2021-02-04 10:49:05.485131       "error" => "vehicle unavailable: {:error=>\"{\\\"error\\\": \\\"timeout\\\"}\"}",
> teslamate_1               | 2021-02-04 10:49:05.485143       "error_description" => "",
> teslamate_1               | 2021-02-04 10:49:05.485148       "response" => nil
> teslamate_1               | 2021-02-04 10:49:05.485152     },
> teslamate_1               | 2021-02-04 10:49:05.485156     headers: [
> teslamate_1               | 2021-02-04 10:49:05.485160       {"date", "Thu, 04 Feb 2021 09:49:05 GMT"},
> teslamate_1               | 2021-02-04 10:49:05.485164       {"content-type", "application/json; charset=utf-8"},
> teslamate_1               | 2021-02-04 10:49:05.485184       {"content-length", "118"},
> teslamate_1               | 2021-02-04 10:49:05.485188       {"connection", "keep-alive"},
> teslamate_1               | 2021-02-04 10:49:05.485192       {"x-frame-options", "SAMEORIGIN"},
> teslamate_1               | 2021-02-04 10:49:05.485195       {"x-xss-protection", "1; mode=block"},
> teslamate_1               | 2021-02-04 10:49:05.485199       {"x-content-type-options", "nosniff"},
> teslamate_1               | 2021-02-04 10:49:05.485202       {"cache-control", "no-cache"},
> teslamate_1               | 2021-02-04 10:49:05.485206       {"x-request-id",
> teslamate_1               | 2021-02-04 10:49:05.485209        "ccde975620b01fb97392c61c912842f7ccde97xxx0b01fb97392c61c912842f7"},
> teslamate_1               | 2021-02-04 10:49:05.485213       {"x-runtime", "9.051839"},
> teslamate_1               | 2021-02-04 10:49:05.485216       {"strict-transport-security", "max-age=15724800; includeSubDomains"}
> teslamate_1               | 2021-02-04 10:49:05.485220     ],
> teslamate_1               | 2021-02-04 10:49:05.485223     method: :get,
> teslamate_1               | 2021-02-04 10:49:05.485227     opts: [
> teslamate_1               | 2021-02-04 10:49:05.485230       access_token: "qts-exxxx1e68305b38a843120xxxxxxxxxxxf8cd538d92317107c4ecf7b2xxxx3c"
> teslamate_1               | 2021-02-04 10:49:05.485235     ],
> teslamate_1               | 2021-02-04 10:49:05.485239     query: [],
> teslamate_1               | 2021-02-04 10:49:05.485242     status: 408,
> teslamate_1               | 2021-02-04 10:49:05.485245     url: "https://owner-api.teslamotors.com/api/1/vehicles/3832xxxxxxxxx2785/vehicle_data"
> teslamate_1               | 2021-02-04 10:49:05.485249   },
> teslamate_1               | 2021-02-04 10:49:05.485252   message: nil,
> teslamate_1               | 2021-02-04 10:49:05.485256   reason: :vehicle_unavailable
> teslamate_1               | 2021-02-04 10:49:05.485259 }
> teslamate_1               | 2021-02-04 10:49:05.485262 2021-02-04 09:49:05.483 [info] Starting logger for 'xxx'
> teslamate_1               | 2021-02-04 10:49:05.492438 2021-02-04 09:49:05.491 [info] Sent 200 in 9123ms
> teslamate_1               | 2021-02-04 10:49:14.608529 2021-02-04 09:49:14.607 car_id=1 [info] Start / :online
> teslamate_1               | 2021-02-04 10:49:14.617152 2021-02-04 09:49:14.616 car_id=1 [info] Connecting ...
> teslamate_1               | 2021-02-04 10:49:26.947338 2021-02-04 09:49:26.946 [info] Connection was closed (a:1|t:1|d:0). Reconnecting …
> ```

Wait, so logging back in worked successfully? The error message here is not unusual, and TeslaMate has no control over that.

**Reply by Maasi1981** [reply]
Adrian I have had the same issue. That means that I have "lost" 6 days of logs.
Am I able to get/fetch those missing data?
I have a gap of the data which is not so good for my statistics.

**Comment by kongomongo** [comment]
@adriankumpf I am not sure if it worked. I guess the token was still valid so everything continued working normal. But I got the token refresh failed with refresh scheduled in 1h => sounds to me refresh failed.

After relogging (2nd post) I get scheduling token refresh in 5 weeks (is this a success message?), and afterwards that error.

I guess it would be better to have an info somewhere if the refresh was successful and/or how long the token is currently valid?
